scripts/anvil_tools/batch_upsert_entities_standard.py

```diff
@@ -0,0 +1,148 @@
+import argparse
+import requests
+
+from oauth2client.client import GoogleCredentials
+
+
+# function to get authorization bearer token for requests
+def get_access_token():
+    """Get access token."""
+
+    scopes = ["https://www.googleapis.com/auth/userinfo.profile", "https://www.googleapis.com/auth/userinfo.email"]
+    credentials = GoogleCredentials.get_application_default()
+    credentials = credentials.create_scoped(scopes)
+
+    return credentials.get_access_token().access_token
+
+
+def call_rawls_batch_upsert(workspace_name, project, request):
+    """Post entities to Terra workspace using batchUpsert."""
+
+    # rawls request URL for batchUpsert
+    uri = f"https://rawls.dsde-prod.broadinstitute.org/api/workspaces/{project}/{workspace_name}/entities/batchUpsert"
+
+    # Get access token and and add to headers for requests.
+    # -H  "accept: */*" -H  "Authorization: Bearer [token] -H "Content-Type: application/json"
+    headers = {"Authorization": "Bearer " + get_access_token(), "accept": "*/*", "Content-Type": "application/json"}
+
+    # capture response from API and parse out status code
+    response = requests.post(uri, headers=headers, data=request)
+    status_code = response.status_code
+
+    if status_code != 204:  # entities upsert fail
+        print(f"WARNING: Failed to upload entities.")
+        print(response.text)
+        return
+
+    # entities upsert success
+    print(f"Successfully uploaded entities." + "\n")
+
+
+def write_request_json(request, filename_prefix):
+    """Create output file with json request."""
+
+    save_name = f"{filename_prefix}_batch_upsert_request.json"
+    with open(save_name, "w") as f:
+        f.write(request)
+
+
+def convert_string_to_list(input_string):
+    """Convert a given string into an array compatible with data model tsvs for array attributes."""
+
+    # remove single & double quotes, remove spaces, remove [ ], separate remaining string on commas (resulting in a list)
+    output_list = str(input_string).replace("'", '').replace('"', '').replace(" ", "").strip('[]').split(",")
+
+    return output_list
+
+
+def create_upsert_request(tsv, array_attr_cols=None):
+    """Generate the request body for batchUpsert API."""
+
+    # check tsv format: data model load tsv requirement "entity:table_name_id" or "membership:table_name_id" -> else exit
```

Comment:
can you add this requirement to the README for input for this function? it only has the workspace/project requirements as far as I can tell

Reply:
It was listed here https://github.com/broadinstitute/horsefish/pull/61/files#diff-19819e317aedad2e9f89b9364f03af1b9ea7bc27949dd41866e299a5d435a37aR151

but i almost missed it myself so i tried to make it more clear to the eye!